asynciterator.js

```diff
@@ -209,24 +220,63 @@ AsyncIterator.prototype.close = function () {
     endAsync(this);
 };
 
+/**
+ Destroy the iterator and stop it from generating new items.
+
+ This will not do anything if the iterator was already ended or destroyed.
+
+ All internal resources will be released an no new items will be emitted,
+ even not already generated items.
+
+ Implementors should not override this method,
+ but instead implement {@link AsyncIterator#_destroy}.
+
+ @param {Error} [error] An optional error to emit.
+ @emits AsyncIterator.end
+ @emits AsyncIterator.error Only emitted if an error is passed.
+ **/
+AsyncIterator.prototype.destroy = function (error) {
+  if (!this.done) {
+    var self = this;
+    this._destroy(error, function (errorInner) {
+      error = error || errorInner;
+      if (error)
+        self.emit('error', error);
+      end(self, true);
+    });
+  }
+};
+
+/**
+ Called by {@link AsyncIterator#destroy}.
+ Implementors can override this, but this should not be called directly.
+
+ @param {Error} error A possible error.
+ @param {Function} callback A callback function with an optional error argument.
+ */
+AsyncIterator.prototype._destroy = function (error, callback) {
+  callback();
+};
+
 /**
   Asynchronously ends the iterator and cleans up.
 
   Should never be called before {@link AsyncIterator#close};
   typically, `close` is responsible for calling `_end`.
 
+  @param {boolean} [destroy] If the iterator should be forcefully destroyed.
   @protected
   @emits AsyncIterator.end
 **/
-AsyncIterator.prototype._end = function () {
-  if (this._changeState(ENDED)) {
+AsyncIterator.prototype._end = function (destroy) {
+  if (this._changeState(destroy ? DESTROYED : ENDED)) {
```

Comment:
Hmm, we should do a careful check of all places that can call `_end`.

Because `changeState` guarantees that we cannot go from a higher state to a lower state, but we could still have the case where `_end` is called on an iterator in state `ENDED`, which would transition it to `DESTROYED`.

Alternatively, we can change `changeState` such that it can never transition beyond `ENDED`.

Reply:
`destroy` has a `if (!this.done)` check, so it won't do anything if the asynciterator is in the `ENDED` state.

I explicitly added a test where `destroy` is called after `close` here: https://github.com/rubensworks/AsyncIterator/blob/d2e1114be1ec2a4dd5fe92146e9c4eac0aa41010/test/AsyncIterator-test.js#L150-L191

Do you think there are other cases that need to be checked/tested as well?

Reply:
Better to be safe than sorry. I made the state transition from `ENDED` to `DESTROYED` impossible in 305304ec3f08f5a62af01ab64c0f0fcfd53dc498.